### DISCUSSION TITLE
Union of path parameter not work

### DISCUSSION THREAD

**Discussion by pyfreyr** [fastapi/fastapi#11551] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from typing import Union, Any


app = FastAPI()

@app.get("/aritfacts/{id_or_name}")
async def get_artifact(
    id_or_name: Union[int, str],
    version: Union[str, None] = None,
) -> Any:
    print(repr(id_or_name))
```

### Description

No matter path param `id_or_name` pass by id(like 86) or by name, router always convert to `str`

```sh
INFO:     Will watch for changes in these directories: ['/export/chenfei167/aaa']
INFO:     Uvicorn running on http://0.0.0.0:3006 (Press CTRL+C to quit)
INFO:     Started reloader process [4132870] using WatchFiles
INFO:     Started server process [4132875]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
c'86'
```

### Operating System

Linux

### Operating System Details

CentOS 7

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.8.5

**Comment by YuriiMotov** [comment, marked as answer]
Hi!
It's a known issue: https://github.com/tiangolo/fastapi/issues/11251

I tried to investigate and it looks like it comes from Pydantic
